lua/lsp_extensions/dart/closing_labels.lua

```diff
@@ -0,0 +1,76 @@
+--[[
+## Closing Labels
+
+** Method: 'dart/textDocument/publishClosingLabels**
+Direction: Server -> Client Params: { uri: string, labels: { label: string, range: Range }[] }
+This notifies the client when closing label information is available (or updated) for a file.
+
+Since this is a notification, the callback needs to be registered in the client's callbacks table.
+This can be achieved with nvim_lspconfig with this minimal config.
+```lua
+nvim_lsp.dartls.setup{
+  init_options = {
+    closingLabels = true,
+  },
+  callbacks = {
+    ['dart/textDocument/publishClosingLabels'] = require('lsp_extensions.dart.closing_labels').get_callback{},
+  },
+}
+```
+
+https://github.com/dart-lang/sdk/blob/master/pkg/analysis_server/tool/lsp_spec/README.md#darttextdocumentpublishclosinglabels-notification
+--]]
+
+local M = {}
+-- Stored labels from the server's notifications keyed by uri
+local all_labels = {}
+-- Namespace for the virtual text
+local closing_labels_ns = vim.api.nvim_create_namespace('lsp_extensions.dart.closing_labels')
+-- Stored opts from the get_callback invocation.
+local label_opts = {}
+
+-- Gets a callback to register to the dartls publishClosingLabels notification.
+-- @tparam table a table of options: highlight, prefix
+M.get_callback = function(opts)
+  label_opts.highlight = opts.highlight or "Comment"
+  label_opts.prefix = opts.prefix or " // "
+
+  vim.cmd [[augroup DartShowClosingLabels]]
```

Comment:
Do we ever / should we ever clear this augroup later?

Reply:
I don't this we would ever want to clean up the autogroup. Maybe if there was functionality to disable the closing labels but that isn't in this PR.